### PR TITLE
MARKET-1468 Multiple Select

### PR DESCRIPTION
**Pull request by Saharij**
Screenshot 2022-12-07 at 02 03 10

### CODE REVIEW

File: components/bl-multiple-select-component/src/options/index.js [hunk 1]
```diff
@@ -23,25 +38,25 @@ export function Options({ type, options, selectValue, setSelectValue, onChange }
       newSelectValue = selectValue.filter(({ objectId }) => objectId !== isOptionSelected.objectId);
     } else {
       const selectedItems = [...selectValue, option];
-      
+
       newSelectValue = type === 'default'
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/index.js:42]: Map select types

File: components/bl-multiple-select-component/src/options/index.js [hunk 2]
```diff
@@ -23,25 +38,25 @@ export function Options({ type, options, selectValue, setSelectValue, onChange }
...
         ? options.filter(item => selectedItems.includes(item))
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/index.js:43]: Here is N^2 complexity. You could create a map of options/selected items and check for existence by looking for the key in the map.

File: components/bl-multiple-select-component/src/options/select-all-checkbox.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { Checkmark } from './checkmark';
+
+export function SelectAllCheckbox({ label, isAllOptionsSelected, handleSelectAll }) {
+  return (
+    <div
+      onClick={ handleSelectAll }
+      className="select-all-checkbox">
+      <Checkmark isOptionSelected={ isAllOptionsSelected } />
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/select-all-checkbox.js:8]:
```
<Checkmark checked={ isAllOptionsSelected } />
```

Checkmark doesn't know about options, isn't it?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-multiple-select-component/src/helpers/validate.js [hunk 4]
```diff
@@ -2,11 +2,11 @@ export const validate = items => {
...
     return [];
   }
-  
+
   if (typeof items === 'string') {
     return items.split(',').map(i => i.trim()).map(i => ({ objectId: i, value: i, label: i }));
   }
-    
+
   if (Array.isArray(items)) {
```
[the remark this answers is not in the document]
Comment (Saharij) [reply components/bl-multiple-select-component/src/helpers/validate.js:10]: We discussed this, but with another component, we don't need the objectId, so I'll remove it.

File: components/bl-multiple-select-component/src/options/select-all-checkbox.js (hunk 3 above)
Comment (Saharij) [reply components/bl-multiple-select-component/src/options/select-all-checkbox.js:8]: Yeap

File: components/bl-multiple-select-component/component.json [hunk 5]
```diff
@@ -26,6 +27,32 @@
       "showInSettings": true,
       "hasLogicHandler": false
     },
+    {
+      "type": "checkbox",
+      "name": "selectAllCheckbox",
+      "label": "Select All Checkbox",
+      "settingTooltip": "This handler allows you to add a checkbox to select all options.",
```
[the remark this answers is not in the document]
Comment (Saharij) [reply components/bl-multiple-select-component/component.json:34]: This is not automatically added, I added it.

File: components/bl-multiple-select-component/src/options/option.js [hunk 6]
```diff
@@ -3,12 +3,11 @@ import { Checkmark } from './checkmark';
 const { cn } = BackendlessUI.CSSUtils;
 
 export function Option({ type, option, isOptionSelected, handleSelectValue }) {
-  
   return (
     <div
       onClick={ () => handleSelectValue(option) }
       className={ cn("option", { "option__selected": isOptionSelected }) }>
-      { type === "checkmark" && <Checkmark isOptionSelected={ isOptionSelected } /> }
+      { type === "checkmark" && <Checkmark checked={ isOptionSelected } /> }
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/option.js:10]: 'checkmark' and 'default' are types of options? Could they be in one object?

File: components/bl-multiple-select-component/src/options/index.js [hunk 7]
```diff
@@ -1,53 +1,66 @@
...
-  
+  }, []);
+
+  const selectedValuesMap = useMemo(() => {
+    return selectValue.reduce((m, { value }) => ({ ...m, [value]: 1 }), {});
+  }, [selectValue]);
+
+  const isAllOptionsSelected = useMemo(() => options.length === Object.keys(selectedValuesMap).length, [selectedValuesMap]);
+
   const handleSelectValue = option => {
```
Comment (v-excelsior) [inline components/bl-multiple-select-component/src/options/index.js:28]: useCallback?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

**Review by Valodya — APPROVED** [review]

**Pull request merged**